File: Host/app/src/main/java/com/demokiller/host/utils/PatchProxy.java
package com.demokiller.host.utils;

import android.text.TextUtils;

import com.demokiller.robustpatch.ChangeQuickRedirect;

public class PatchProxy {
	
    public static boolean isSupport(Object[] argsObj, Object thisObj, ChangeQuickRedirect changeQuickRedirect, boolean isStatic) {
        if (changeQuickRedirect == null) {
            return false;
        }
        String classMethod = getClassMethod(isStatic);
        if (TextUtils.isEmpty(classMethod)) {
            return false;
        }
        return changeQuickRedirect.isSupport(classMethod, getObjects(argsObj, thisObj, isStatic));
    }

    public static Object accessDispatch(Object[] argsObj, Object thisObj, ChangeQuickRedirect changeQuickRedirect, boolean isStatic) {
        if (changeQuickRedirect == null) {
            return null;
        }
        String classMethod = getClassMethod(isStatic);
        if (TextUtils.isEmpty(classMethod)) {
            return null;
        }
        return changeQuickRedirect.accessDispatch(classMethod, getObjects(argsObj, thisObj, isStatic));
    }

    public static void accessDispatchVoid(Object[] argsObj, Object thisObj, ChangeQuickRedirect changeQuickRedirect, boolean isStatic) {
        if (changeQuickRedirect != null) {
            String classMethod = getClassMethod(isStatic);
            if (!TextUtils.isEmpty(classMethod)) {
                changeQuickRedirect.accessDispatch(classMethod, getObjects(argsObj, thisObj, isStatic));
            }
        }
    }

    private static Object[] getObjects(Object[] argsObj, Object thisObj, boolean isStatic) {
        if (argsObj == null) {
            return null;
        }
        Object[] newArgsObj;
        int length = argsObj.length;
        if (isStatic) {
        	newArgsObj = new Object[length];
        } else {
        	newArgsObj = new Object[length + 1];
        }
        int i = 0;
        while (i < length) {
        	newArgsObj[i] = argsObj[i];
            i++;
        }
        if (isStatic) {
            return newArgsObj;
        }
        newArgsObj[i] = thisObj;
        return newArgsObj;
    }

    private static String getClassMethod(boolean isStatic) {
        String str = "";
        try {
            StackTraceElement stackTraceElement = new Throwable().getStackTrace()[2];
            str = stackTraceElement.getClassName() + ":" + stackTraceElement.getMethodName() + ":" + isStatic;
        } catch (Throwable th) {
        }
        return str;
    }
}

File: Host/app/src/main/java/com/demokiller/host/utils/PatchProxy.kt
package com.demokiller.host.utils

import android.text.TextUtils
import com.demokiller.robustpatch.ChangeQuickRedirect

object PatchProxy {
    fun isSupport(argsObj: Array<Any>?, thisObj: Any?, changeQuickRedirect: ChangeQuickRedirect?, isStatic: Boolean): Boolean {
        if (changeQuickRedirect == null) {
            return false
        }
        val classMethod = getClassMethod(isStatic)
        return if (TextUtils.isEmpty(classMethod)) {
            false
        } else changeQuickRedirect.isSupport(classMethod, getObjects(argsObj, thisObj, isStatic))
    }

    fun accessDispatch(argsObj: Array<Any>?, thisObj: Any?, changeQuickRedirect: ChangeQuickRedirect?, isStatic: Boolean): Any? {
        if (changeQuickRedirect == null) {
            return null
        }
        val classMethod = getClassMethod(isStatic)
        return if (TextUtils.isEmpty(classMethod)) {
            null
        } else changeQuickRedirect.accessDispatch(classMethod, getObjects(argsObj, thisObj, isStatic))
    }

    fun accessDispatchVoid(argsObj: Array<Any>?, thisObj: Any, changeQuickRedirect: ChangeQuickRedirect?, isStatic: Boolean) {
        if (changeQuickRedirect != null) {
            val classMethod = getClassMethod(isStatic)
            if (!TextUtils.isEmpty(classMethod)) {
                changeQuickRedirect.accessDispatch(classMethod, getObjects(argsObj, thisObj, isStatic))
            }
        }
    }

    private fun getObjects(argsObj: Array<Any>?, thisObj: Any?, isStatic: Boolean): Array<Any?>? {
        if (argsObj == null) {
            return null
        }
        val newArgsObj: Array<Any?>
        val length = argsObj.size
        newArgsObj = if (isStatic) {
            arrayOfNulls(length)
        } else {
            arrayOfNulls(length + 1)
        }
        var i = 0
        while (i < length) {
            newArgsObj[i] = argsObj[i]
            i++
        }
        if (isStatic) {
            return newArgsObj
        }
        newArgsObj[i] = thisObj
        return newArgsObj
    }

    private fun getClassMethod(isStatic: Boolean): String {
        var str = ""
        try {
            val stackTraceElement = Throwable().stackTrace[2]
            str = stackTraceElement.className + ":" + stackTraceElement.methodName + ":" + isStatic
        } catch (th: Throwable) {
        }
        return str
    }
}